syft/serde/protobuf/serde.py

```diff
@@ -0,0 +1,355 @@
+    # otherwise we output the compressed stream with header set to '1'
+    # even if compressed flag is set to false by the caller we
+    # output the input stream as it is with header set to '0'
+    if force_no_compression:
+        return binary
+    else:
+        return compression._compress(binary)
+
+
+def deserialize(binary: bin, worker: AbstractWorker = None, unbufferizes=True) -> object:
+    """ This method can deserialize any object PySyft needs to send or store.
+
+    This is the high level function for deserializing any object or collection
+    of objects which PySyft has sent over the wire or stored. It includes three
+    steps, Decompress, Deserialize, and Detail as described inline below.
+
+    Args:
+        binary (bin): the serialized object to be deserialized.
+        worker (AbstractWorker): the worker which is acquiring the message content,
+            for example used to specify the owner of a tensor received(not obvious
+            for virtual workers)
+        unbufferizes (bool): there are some cases where we need to perform the decompression
+            and deserialization part, but we don't need to unbufferize all the message.
+            This is the case for Plan workers for instance
+
+    Returns:
+        object: the deserialized form of the binary input.
+    """
+    if worker is None:
+        # TODO[jvmancuso]: This might be worth a standalone function.
+        worker = syft.framework.hook.local_worker
+
+    # 1) Decompress the binary if needed
+    binary = compression._decompress(binary)
+
+    # 2) Deserialize
+    msg_wrapper = SyftMessagePB()
+    msg_wrapper.ParseFromString(binary)
+
+    # 3) Convert back to a Python object
+    message_type = msg_wrapper.WhichOneof("contents")
+    python_obj = _unbufferize(worker, getattr(msg_wrapper, message_type))
+    return python_obj
+
+
+def create_protobuf_id(id) -> IdPB:
+    protobuf_id = IdPB()
+    if type(id) == type("str"):
+        protobuf_id.id_str = id
+    else:
+        protobuf_id.id_int = id
+    return protobuf_id
+
+
+def _bufferize(worker: AbstractWorker, obj: object, **kwargs) -> object:
+    """
+    This function takes an object as input and returns a
+    Protobuf object. The reason we have this function
+    is that some objects are either NOT supported by high level (fast)
+    serializers OR the high level serializers don't support the fastest
+    form of serialization. For example, PyTorch tensors have custom pickle
+    functionality thus its better to pre-serialize PyTorch tensors using
+    pickle and then serialize the binary in with the rest of the message
+    being sent.
+
+    Args:
+        obj: An object which needs to be converted to Protobuf.
+
+    Returns:
+        An Protobuf object which Protobuf can serialize.
+    """
+
+    # Check to see if there is a bufferizer
+    # for this type. If there is, return the bufferized object.
+    # breakpoint()
+    current_type = type(obj)
```

Comment:
I think I saw kwargs passed into this func. Will it handle dict?

Reply:
Not yet; it only handles [a tiny number of classes](https://github.com/OpenMined/PySyft/pull/2826/files/b0e11eb0d61813eac5e3ba8271004617957bc387#diff-24eeb7061c1f9cb3bdcc09732fb58acbR40) right now. I haven't ironed out exactly how Python primitives are going to be handled, but hope that it can be similar to the way Torch types get serialized.

Reply:
Kwargs are working in #2877. This function doesn't end up receiving a `dict`, so it works out okay.